src/ast/pass_manager.cpp

```diff
@@ -1,47 +1,75 @@
 #include "ast/pass_manager.h"
-
-#include <ostream>
-
-#include "ast/passes/printer.h"
-#include "bpftrace.h"
+#include "ast/context.h"
+#include "log.h"
+#include "util/result.h"
 
 namespace bpftrace::ast {
 
-namespace {
-void print(PassContext &ctx, const std::string &name, std::ostream &out)
+std::atomic<int> PassContext::next_type_id_;
+std::unordered_map<int, std::string> PassContext::type_names_;
+
+bool PassContext::ok() const
 {
-  out << "\nAST after: " << name << std::endl;
-  out << "-------------------\n";
-  ast::Printer printer(out);
-  printer.visit(ctx.ast_ctx.root);
-  out << std::endl;
+  int type_id = TypeId<ASTContext>::type_id();
+  if (state_.contains(type_id)) {
+    return get<ASTContext>().diagnostics().ok();
+  }
+  return true;
 }
-} // namespace
 
-void PassManager::AddPass(Pass p)
+void PassContext::fail(int type_id)
```

Comment:
this fail seems a bit specific - what about re-naming to something like `no_object_fail` or just inlining this logic where it's called.

Reply:
Done.